test/vesting.js

```diff
@@ -0,0 +1,347 @@
+const { expect, assert } = require("chai");
+const { ethers, upgrades } = require("hardhat");
+
+describe("Vesting Contracts", function () {
+  describe("Single Account Vesting", function () {
+    let arrowToken;
+    let vestingFactory;
+    let vestingContract;
+    let beneficiary1;
+
+    let startTimestamp;
+    let durationSeconds = 100;
+    let amount = 1_000_000;
+
+    it("Should correctly vest", async function () {
+      let VestingFactory = await ethers.getContractFactory(
+        "ArrowVestingFactory"
+      );
+      let Token = await ethers.getContractFactory("ArrowToken");
+      let name = "Arrow Token";
+      let symbol = "ARROW";
+      arrowToken = await Token.deploy(
+        ethers.BigNumber.from(1_000_000),
+        name,
+        symbol
+      );
+      await arrowToken.deployed();
+      let VestingBase = await ethers.getContractFactory("ArrowVestingBase");
+
+      [_, beneficiary1] = await ethers.getSigners();
+      vestingFactory = await VestingFactory.deploy();
+      await vestingFactory.deployed();
+
+      // get block timestamp
+      const blockNum = await ethers.provider.getBlockNumber();
+      const block = await ethers.provider.getBlock(blockNum);
+      startTimestamp = block.timestamp + 100;
+
+      // use callStatic to get what the address would be when the transaction is created
+      let vestingWalletAddress =
+        await vestingFactory.callStatic.createVestingSchedule(
+          beneficiary1.address,
+          startTimestamp,
+          durationSeconds
+        );
+      // actually create the transaction so that the child contract can be deployed on chain
+      let txn = await vestingFactory.createVestingSchedule(
+        beneficiary1.address,
+        startTimestamp,
+        durationSeconds
+      );
+      await txn.wait();
+      vestingContract = await VestingBase.attach(vestingWalletAddress);
+
+      /*---------------------------------Test Begin---------------------------------**/
+
+      // Add tokens to the contract
+      await arrowToken.transfer(vestingContract.address, amount);
+      expect(await arrowToken.balanceOf(vestingContract.address)).to.equal(
+        amount
+      );
+      console.log("√ Pass - Add tokens to the contract");
+
+      // Check contract variables have been set correctly
+      expect(await vestingContract.start()).to.equal(startTimestamp);
+      expect(await vestingContract.duration()).to.equal(durationSeconds);
+      expect(
+        await vestingContract
+          .connect(beneficiary1)
+          ["released(address)"](arrowToken.address)
+      ).to.equal(0);
+      console.log("√ Pass - Check contract variables have been set correctly");
+
+      //No tokens should have been immediately released
+      expect(await arrowToken.balanceOf(beneficiary1.address)).to.equal(0);
+      await vestingContract
+        .connect(beneficiary1)
+        ["release(address)"](arrowToken.address);
```

Comment:
```suggestion
       .release(arrowToken.address);
```
Why this notation? I'm unfamiliar with hardhat, but can't we just make a normal function call rather than calling into a map?

Reply:
This is because the OpenZeppelin [lib](https://docs.openzeppelin.com/contracts/4.x/api/finance#VestingWallet) has two `release` functions. Hardhat cannot automatically handle the overload, so this particular notation is required to let hardhat know which `release` we are calling. Otherwise, there will be an error when running the code.

https://ethereum.stackexchange.com/questions/99904/hardhat-cant-find-the-mint-function

Reply:
Ah good to know. Thanks.